common/raw_hashtable.h

```diff
@@ -0,0 +1,1302 @@
+  // tempting to reuse the already allocated storage, but doing so appears to
+  // be a (very slight) performance regression. These are relatively rare and
+  // storing them into the existing storage creates stores to the same regions
+  // of memory we're reading. Moreover, it requires moving both the key and the
+  // value twice, and doing the `memcpy` widening for relocatable types before
+  // the group walk rather than after the group walk. In practice, between the
+  // statistical rareness and using a large small size buffer here on the stack,
+  // we can handle this most efficiently with temporary, additional storage.
+  llvm::SmallVector<ssize_t, 128> probed_indices;
+
+  // We grow into a new `MapBase` so that both the new and old maps are
+  // fully functional until all the entries are moved over. However, we directly
+  // manipulate the internals to short circuit many aspects of the growth.
+  ssize_t old_size = alloc_size();
+  CARBON_DCHECK(old_size > 0);
+  CARBON_DCHECK(growth_budget_ == 0);
+
+  bool old_small = is_small();
+  Storage* old_storage = storage();
+  uint8_t* old_metadata = metadata();
+  EntryT* old_entries = entries();
+
+#ifndef NDEBUG
+  ssize_t debug_empty_count =
+      llvm::count(llvm::ArrayRef(old_metadata, old_size),
+                  MetadataGroup::Empty) +
+      llvm::count(llvm::ArrayRef(old_metadata, old_size),
+                  MetadataGroup::Deleted);
+  CARBON_DCHECK(debug_empty_count >=
+                (old_size - GrowthThresholdForAllocSize(old_size)))
+      << "debug_empty_count: " << debug_empty_count << ", size: " << old_size;
+#endif
+
+  // Compute the new size and grow the storage in place (if possible).
+  ssize_t new_size = ComputeNextAllocSize(old_size);
+  alloc_size() = new_size;
+  storage() = Allocate(new_size);
+  growth_budget_ = GrowthThresholdForAllocSize(new_size);
+
+  // Now extract the new components of the table.
+  uint8_t* new_metadata = metadata();
+  EntryT* new_entries = entries();
+
+  // The common case, especially for large sizes, is that we double the size
+  // when we grow. This allows an important optimization -- we're adding
+  // exactly one more high bit to the hash-computed index for each entry. This
+  // in turn means we can classify every entry in the table into three cases:
+  //
+  // 1) The new high bit is zero, the entry is at the same index in the new
+  //    table as the old.
+  //
+  // 2) The new high bit is one, the entry is at the old index plus the old
+  //    size.
+  //
+  // 3) The entry's current index doesn't match the initial hash index because
+  //    it required some amount of probing to find an empty slot.
+  //
+  // The design of the hash table is specifically to minimize how many entries
+  // fall into case (3), so we expect the vast majority of entries to be in
+  // (1) or (2). This lets us model growth notionally as duplicating the hash
+  // table up by `old_size` bytes, clearing out the empty slots, and inserting
```

Comment:
`old_size` is measuring slots not bytes, so this seems weird. Maybe:

```suggestion
  // table metadata by `old_size` bytes, clearing out the empty slots, and inserting
```
or
```suggestion
  // table, clearing out the empty slots, and inserting
```